### DISCUSSION TITLE
Failed verification: ' fStageListeners.empty() ' -- Coding Error in onStageSet at line 517

### DISCUSSION THREAD

**Discussion by csyshing** [Autodesk/maya-usd#2074] [Q&A] [answered]
Hi,

I am debugging an error from a render farm job that was causing by this message in `maya-usd`:
```
Failed verification: ' fStageListeners.empty() ' -- Coding Error in onStageSet at line 517
```
Searching around the code and seems this variable `fStageListeners` has been used only in `StagesSubject` class, and it's not clear to me how come the verification failed at line 517?
https://github.com/Autodesk/maya-usd/blob/dac1abbd24282b45ef5be7ed288f2180142dee20/lib/mayaUsd/ufe/StagesSubject.cpp#L516-L517

by looking at the code, line 517 doesn't seem to affect anything else, the code still keep going forward regardless of the verification - is this severity level too high? Can we maybe lower it to be a warning instead?
Or if `fStageListeners` is critical, how about adding proper error handling / clean up?

Thanks,
Zhicheng

**Comment by kxl-adsk** [comment, marked as answer]
This is important to investigate issue. We are sending onStageInvalidate from dirty propagation, and after onStageSet from compute when the new stage is set. The error is telling us that something refreshed the map of all stages but with invalid data. This means that some consumers of this information (either read directly for proxy shape or via stage cache) have received incorrect information.

Most likely, something causes pull evaluation during dirty propagation. This is illegal in Maya's DG.

**Reply by csyshing** [reply]
Thanks @kxl-adsk , I will do some investigation for the consumer code then.
